storage_controller/src/service.rs

```diff
@@ -2376,61 +2376,75 @@ impl Service {
         let _tenant_lock =
             trace_exclusive_lock(&self.tenant_op_locks, tenant_id, TenantOperations::Delete).await;
 
-        self.ensure_attached_wait(tenant_id).await?;
-
-        // TODO: refactor into helper
-        let targets = {
-            let locked = self.inner.read().unwrap();
-            let mut targets = Vec::new();
-
+        // Detach all shards
+        let (detach_waiters, shard_ids, node) = {
+            let mut shard_ids = Vec::new();
+            let mut detach_waiters = Vec::new();
+            let mut locked = self.inner.write().unwrap();
+            let (nodes, tenants, scheduler) = locked.parts_mut();
             for (tenant_shard_id, shard) in
-                locked.tenants.range(TenantShardId::tenant_range(tenant_id))
+                tenants.range_mut(TenantShardId::tenant_range(tenant_id))
             {
-                let node_id = shard.intent.get_attached().ok_or_else(|| {
-                    ApiError::InternalServerError(anyhow::anyhow!("Shard not scheduled"))
-                })?;
-                let node = locked
-                    .nodes
-                    .get(&node_id)
-                    .expect("Pageservers may not be deleted while referenced");
+                shard_ids.push(*tenant_shard_id);
 
-                targets.push((*tenant_shard_id, node.clone()));
+                shard.set_scheduling_policy(ShardSchedulingPolicy::Pause);
+                shard.policy = PlacementPolicy::Detached;
+                shard.intent.clear(scheduler);
+                if let Some(waiter) = self.maybe_reconcile_shard(shard, nodes) {
+                    detach_waiters.push(waiter);
+                }
```

Comment:
This reads beatifully and correct, but honestly I don't have all the invariants of the storage controller in my head.

One question that came to mind: what if there's a reconciliation going on while we make these updates to the desired state (scheduling_policy, placement policy, intent).

It seems to me that `maybe_reconcile_shard` returns the waiter for that ongoing reconciliation.

But does that ongoing reconciliation notice that we changed the intended state from underneath it?

Reply:
> One question that came to mind: what if there's a reconciliation going on while we make these updates to the desired state (scheduling_policy, placement policy, intent).
> It seems to me that maybe_reconcile_shard returns the waiter for that ongoing reconciliation.

Ah: you've spotted a mistake.  maybe_reconcile_shard only returns a WaitExisting if the sequence number on the tenant hasn't advanced since the current reconciliation... but when I directly set the intent here I'm not bumping the sequence number (whereas TenantShard::schedule does update sequence when it modifies anything)

This code should have gone via `schedule()` to get the proper behavior -- fixed in 55f2dd5a7

Later, we should make `sequence` and `intent` private so that the "sequence increases when we modify intent" behavior is better encapsulated.  TenantShard is a bit leakier than it should be.